app/routes/home.search.$searchTerm.tsx

```diff
@@ -1,81 +1,275 @@
+      throw new Error(getSavedArticlesResponse.message, {
+        cause: getSavedArticlesResponse,
+      })
+    }
+
+    return json({
+      signedIn: true,
+      articles,
+      savedArticles: getSavedArticlesResponse.data,
+      amount,
+    })
+  }
+
+  return json({
+    signedIn: false,
+    articles,
+    savedArticles: null,
+    amount,
+  })
 }
 
-export function ErrorBoundary() {
-  const error = useRouteError()
+export const action = async ({
+  request,
+}: ActionFunctionArgs): ActionReturnType => {
+  const { session, authState } = await serverAuthPublicRoute(request)
 
-  if (isRouteErrorResponse(error)) {
-    return (
-      <div>
-        <h1>Oops</h1>
-        <p>Status: {error.status}</p>
-        <p>{error.data.message}</p>
-      </div>
-    )
+  if (authState === AuthState.tokenValidationFailed) {
+    return redirect(Route.home, {
+      headers: { 'Set-Cookie': await destroySession(session) },
+    })
+  }
+
+  const formData = await request.formData()
+  const serializedArticle = formData.get('article')
+  invariant(
+    serializedArticle && typeof serializedArticle === 'string',
+    'Missing or invalid article',
+  )
+  const article = JSON.parse(serializedArticle)
+
+  const { success, response } = await saveArticle(article, session)
+
+  if (!success) {
+    console.error('failed to save article')
+    console.error(response)
+    return json({ success: false, message: 'failed' }, { status: 500 })
   }
 
-  return <NoArticle />
+  console.log('save article success', article)
+
+  return json({ success: true, message: 'success' }, { status: 200 })
 }
 
 export default function SearchResults() {
-  const { response: articles, amount } = useLoaderData<typeof loader>()
+  const { signedIn, articles, savedArticles, amount } =
+    useLoaderData<typeof loader>()
   const resultsRef = useRef<HTMLDivElement>(null)
+  const navigation = useNavigation()
 
   useEffect(() => {
-    if (amount === '6') {
+    const fromSearch = navigation.location?.state?.fromSearch
+    if (fromSearch) {
       resultsRef.current?.scrollIntoView({ behavior: 'smooth' })
     }
-  }, [articles, amount])
+  }, [navigation])
 
-  return (
-    <Suspense fallback={<Loading />}>
+  return articles.length > 0 ? (
+    <>
       <ArticleGalleryLayout
         title='Search results'
         topRef={resultsRef}
-        amount={amount}
+        amountParam={amount}
+        amount={articles.length}
       >
-        {articles.map((article) => (
-          <ResultArticleCard data={article} key={article._id} />
-        ))}
+        {articles.slice(0, amount).map((article) => {
+          let isSaved = false
+          if (signedIn) {
+            isSaved = savedArticles.some(
+              (savedArticle) => savedArticle.link === article.url,
+            )
+          }
+          return (
+            <ResultArticleCard
+              isSaved={isSaved}
+              data={article}
+              key={Math.random()}
```

Comment:
news api articles dont have an id, i can use the url though.

Reply:
Regarding the context provider: there is no reason for me to use **client** global state in this application since its a server rendered application, the state is always updated since navigations cause loaders to run and update the data before the page is displayed.

The global state is the server state